src/pydsl/scf.py

```diff
@@ -2,12 +2,299 @@
 
 from mlir.dialects import scf
 from mlir.ir import InsertionPoint
+import mlir.ir as mlirir
 
 from pydsl.macro import Compiled, IteratorMacro, Uncompiled
-from pydsl.protocols import Lowerable, SubtreeOut, ToMLIRBase
+from pydsl.protocols import (
+    CompileTimeTestable,
+    ToMLIRBase,
+)
+from pydsl.scope import BindingContext, Scope
 from pydsl.type import Index, lower_single
 
 
+def build_block(
+    visitor: ToMLIRBase,
+    scope: Scope,
+    block_body: list[ast.AST],
+    dest_block: mlirir.Block,
+    node_for_context: ast.AST,
+    assign_targets: dict[str, mlirir.Value] | None = None,
+):
+    """
+    Compile an AST block in a new binding context and return the updated local
+    bindings.
+
+    Parameters
+    ----------
+    visitor : ToMLIRBase
+        The visitor used to compile AST nodes.
+    scope : Scope
+        The current scope.
+    block_body : list[ast.AST]
+        The list of AST nodes (statements) to compile.
+    dest_block : mlir.ir.Block
+        The MLIR block to insert ops into.
+    binding_node : ast.AST
+        The AST node that owns the binding context.
+    assign_targets : dict[str, mlirir.Value] | None
+        Optional dict of names to assign before executing the block (e.g., induction variable).
+
+    Returns
+    -------
+    dict[str, mlirir.Value]
+        new bindings after the block.
+    """
+
+    binding_prior = scope.f_locals
+
+    with (
+        scope.new_binding_context(BindingContext(node_for_context, dict())),
```

Comment:
The only place we ever use this context manager is here. Why can we just manually call `scope.stack.append(BindingContext(...))` here? it seems like it is mainly to avoid the exception issues, which should be handled in a different way.